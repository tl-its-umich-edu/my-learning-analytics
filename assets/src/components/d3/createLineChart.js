import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { margin } from '../../constants/chartConstants'
import createSVG from './createSVG'
import createLine from './createLine'

function createLineChart ({ data, width, height, el }) {
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  const x = d3.scaleLinear()
    .domain([0, d3.max(data, d => d.label)]).nice()
    .range([margin.left, aWidth - margin.right])

  const y = d3.scaleLinear()
    .domain([0, d3.max(data, d => d.data)]).nice()
    .range([aHeight - margin.bottom, margin.top])

  const xAxis = g => g
    .attr('transform', `translate(0,${aHeight - margin.bottom})`)
    .call(d3.axisBottom(x).ticks(aWidth / 80).tickSizeOuter(0))

  const yAxis = g => g
    .attr('transform', `translate(${margin.left},0)`)
    .call(d3.axisLeft(y))

  const svg = createSVG({ el, width, height })

  createLine({ svg, x, y, data })

  svg.append('g')
    .call(xAxis)

  svg.append('g')
    .call(yAxis)
}

export default createLineChart
