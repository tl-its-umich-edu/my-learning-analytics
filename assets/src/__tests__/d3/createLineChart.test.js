/* global describe, test, expect */
import createLineChart from '../../components/d3/createLineChart'

const lineChartData = Object.freeze(
  [
    {
      'label': 0,
      'data': 0
    },
    {
      'label': 1,
      'data': 1
    },
    {
      'label': 2,
      'data': 2
    },
    {
      'label': 3,
      'data': 2
    },
    {
      'label': 4,
      'data': 3
    }
  ]
)

describe('createLineChart', () => {
  test('should build a line chart', () => {
    const div = document.createElement('div')
    createLineChart({ data: lineChartData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
