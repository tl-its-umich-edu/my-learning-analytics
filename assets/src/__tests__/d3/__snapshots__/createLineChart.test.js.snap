// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLineChart should build a line chart 1`] = `
<div>
  <svg
    height="450"
    width="940"
  >
    <path
      d="M40,420L260,286.6666666666667L480,153.33333333333337L700,153.33333333333337L920,20"
      fill="none"
      stroke="steelblue"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.5"
    />
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0,420)"
    >
      <path
        class="domain"
        d="M40.5,0.5H920.5"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(150.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(260.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(370.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          1.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(480.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          2.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(590.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          2.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(700.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          3.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(810.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          3.5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(920.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          4.0
        </text>
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,420.5H0.5V20.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,420.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,393.8333333333333)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0.2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,367.1666666666667)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0.4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,340.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0.6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,313.8333333333333)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0.8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,287.1666666666667)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,260.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1.2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,233.83333333333334)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1.4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,207.16666666666666)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1.6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,180.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1.8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,153.83333333333337)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2.0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,127.16666666666663)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2.2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,100.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2.4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,73.83333333333331)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2.6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,47.166666666666686)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2.8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,20.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          3.0
        </text>
      </g>
    </g>
  </svg>
</div>
`;
